sql/sem: reject AS OF timestamps set in the future

spilchen opened the pull request:
Recent support for AS OF timestamps in CTAS and MQTs exposed inconsistent behavior when the timestamp is set in the future:
- In some cases, operations fail with a KV read error.
- In others, they pause briefly before returning.
- When the timestamp is significantly in the future, operations may hang indefinitely.

 This behaviour was observed in issue #145781, where a future AS OF timestamp caused TestRandomSyntaxGeneration to fail due to a hang.

This change ensures AS OF timestamps are not in the future:
- For computed timestamps (e.g., using intervals), enforcement is strict.
- For fixed timestamps, a tolerance is allowed to accommodate clock skew and other cases.

One exception remains: AS OF timestamps are still allowed in the future when creating a CHANGEFEED, as its end_time parameter is expected to be in the future and flows through as an AS OF time.

Fixes #145781

Epic: none
Release note (bug fix): Prevent use of future AS OF timestamps in CTAS and materialized views. Previously, such timestamps could cause hangs.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146446)
<!-- Reviewable:end -->

spilchen committed:
sql/sem: reject AS OF timestamps set in the future

Recent support for AS OF timestamps in CTAS and MQTs exposed
inconsistent behavior when the timestamp is set in the future:
- In some cases, operations fail with a KV read error.
- In others, they pause briefly before returning.
- When the timestamp is significantly in the future, operations
  may hang indefinitely.

This behaviour was observed in issue #145781, where a future AS OF
timestamp caused TestRandomSyntaxGeneration to fail due to a hang.

This change ensures AS OF timestamps are not in the future:
- For computed timestamps (e.g., using intervals), enforcement is strict.
- For fixed timestamps, a tolerance is allowed to accommodate clock skew
  and other cases.

One exception remains: AS OF timestamps are still allowed in the future
when creating a CHANGEFEED, as its end_time parameter is expected to be
in the future and flows through as an AS OF time.

Fixes #145781

Epic: none
Release note (bug fix): Prevent use of future AS OF timestamps in CTAS and
materialized views. Previously, such timestamps could cause errors, delays, or
hangs.

fqazi reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Just one minor issue for this one

> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146446#-:-OQ48yrjCRyq3I73IhuE:bd4usbu)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aerfrei)

___
*[`pkg/sql/sem/asof/as_of.go` line 368 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146446#-OQ48dF1AQmUJjxWCqdE:-OQ48dF1AQmUJjxWCqdF:bknid1c) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d06398926f5c5b7970c7658b10bee22fa0dfe58/pkg/sql/sem/asof/as_of.go#L368)):*
> ```Go
> 
> 		if ts.WallTime > maxAllowedWallTime {
> 			return ts, errors.Errorf("timestamp %v is in the future", d)
> ```

A pgerror might be better here since this is supposed to be surface to users.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
since this is fixing a hang in the MATERIALIZED VIEW statements, let's backport to 25.2

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146446#-:-OQ4OCzxAXbjS9raX06d:b2zj6km)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @aerfrei and @fqazi)

___
*[`pkg/sql/sem/asof/as_of.go` line 368 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146446#-OQ48dF1AQmUJjxWCqdE:-OQ4NVGUDq1_It-47Zo8:b-a268ij) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d06398926f5c5b7970c7658b10bee22fa0dfe58/pkg/sql/sem/asof/as_of.go#L368)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

A pgerror might be better here since this is supposed to be surface to users.
</blockquote></details>

+1, i think the same should apply to all the other errors in this function too. 

could you add a separate commit to to use `pgerror.Newf(pgcode.InvalidParameterValue, ...)` for these errors?


<!-- Sent from Reviewable.io -->

spilchen committed:
sql/sem: convert DatumToHLC error reporting to pgerror

The DatumToHLC function previously surfaced client-facing errors using
simple errors.Errorf. This commit updates all such code paths to
use pgerror.Newf instead. This change allows for attaching appropriate
pgcodes to errors.

Epic: none
Release note: none

spilchen commented on pkg/sql/sem/asof/as_of.go:
```diff
@@ -253,6 +265,9 @@ const (
 	// ALTER VIRTUAL CLUSTER ... COMPLETE REPLICATION statement.
 	ReplicationCutover
 
+	// AsOfFuture is like AsOf, except the timestamp is expected to be in the future.
```
Yes, good point. Let me update the comment.
<!-- Reviewable comment -OQ579TDApCiSgChaaKj:b-awa7tl -->
<!-- Sent from Reviewable.io -->

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146446)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aerfrei and @fqazi)

___
*[`pkg/sql/sem/asof/as_of.go` line 368 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/146446#-OQ48dF1AQmUJjxWCqdE:-OQ578sw3Js3LudWDy5A:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d06398926f5c5b7970c7658b10bee22fa0dfe58/pkg/sql/sem/asof/as_of.go#L368)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

+1, i think the same should apply to all the other errors in this function too. 

could you add a separate commit to to use `pgerror.Newf(pgcode.InvalidParameterValue, ...)` for these errors?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

spilchen commented:
TFTRs!

bors r+
